fix: aws session creation is failing for s3 manager when roles are used

koladilip opened the pull request:
# Description

Unify AWS session config creation process so that it is consistent across all AWS destinations. Earlier S3 has slightly different process to create SessionConfing so when updated the session config creation in aws utils to support RoleBasedAuth flag, it got missed for S3 manager as it was using different implementation so this unification should prevent such errors in the future.

## Notion Ticket

[Warehouse AWS Role support] https://www.notion.so/rudderstacks/Warehouse-destination-config-for-Role-base-Authentication-control-plane-db9198dce2024723959c224e91a11ef4
## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

crownedlake reviewed (approved):
LGTM.

atzoum commented:
@koladilip what are we fixing and what issue did this cause? [ref](https://rudderlabs.slack.com/archives/C02B31USATA/p1670577005889579)

atzoum commented on services/filemanager/s3manager_test.go:
```diff
@@ -26,44 +30,17 @@ func TestGetSessionWithAccessKeys(t *testing.T) {
 	assert.NotNil(t, s3Manager.session)
 }
-
-func TestGetSessionConfigWithAccessKeys(t *testing.T) {
-	s3Manager := S3Manager{
-		Config: &S3Config{
-			Bucket:      "someBucket",
-			AccessKeyID: "someAccessKeyId",
-			AccessKey:   "someSecretAccessKey",
-			Region:      aws.String("someRegion"),
-		},
-	}
-	awsSessionConfig := s3Manager.getSessionConfig()
-	assert.NotNil(t, awsSessionConfig)
-	assert.Equal(t, s3Manager.Config.AccessKey, awsSessionConfig.AccessKey)
-	assert.Equal(t, s3Manager.Config.AccessKeyID, awsSessionConfig.AccessKeyID)
-}
-
-func TestGetSessionConfigWithIAMRole(t *testing.T) {
-	s3Manager := S3Manager{
-		Config: &S3Config{
-			Bucket:     "someBucket",
-			IAMRoleARN: "someIAMRole",
-			ExternalID: "someExternalID",
-			Region:     aws.String("someRegion"),
-		},
-	}
-	awsSessionConfig := s3Manager.getSessionConfig()
-	assert.NotNil(t, awsSessionConfig)
-	assert.Equal(t, s3Manager.Config.IAMRoleARN, awsSessionConfig.IAMRoleARN)
-	assert.Equal(t, s3Manager.Config.ExternalID, awsSessionConfig.ExternalID)
-}
```
Aren't these meaningful anymore? Are we introducing any other tests to verify that:

1. what we are fixing is actually fixed 
2. no regression has been introduced in the meantime by these changes

koladilip replied:
Yes `s3Manager.getSessionConfig()` is removed so I have added few more tests to test these functionalities.

koladilip commented:
> @koladilip what are we fixing and what issue did this cause? [ref](https://rudderlabs.slack.com/archives/C02B31USATA/p1670577005889579)

@atzoum 
Since S3 manager was using different implementation, when we refactored Session config creation in awsutils, we forgot the S3 manager and it introduced the bug so unifying all of them to use same implementation should fix the bug.

atzoum commented:
> > what are we fixing and what issue did this cause? [ref](https://rudderlabs.slack.com/archives/C02B31USATA/p1670577005889579)
> 
> @atzoum Since S3 manager was using different implementation, when we refactored Session config creation in awsutils, we forgot the S3 manager and it introduced the bug so unifying all of them to use same implementation should fix the bug.

What was the issue this bug caused? Can you please include it in the pull request's title?

koladilip commented:
> > > what are we fixing and what issue did this cause? [ref](https://rudderlabs.slack.com/archives/C02B31USATA/p1670577005889579)
> > 
> > 
> > @atzoum Since S3 manager was using different implementation, when we refactored Session config creation in awsutils, we forgot the S3 manager and it introduced the bug so unifying all of them to use same implementation should fix the bug.
> 
> What was the issue this bug caused? Can you please include it in the pull request's title?

Updated the PR title, let me know this explains the problem.

atzoum commented:
> > > > what are we fixing and what issue did this cause? [ref](https://rudderlabs.slack.com/archives/C02B31USATA/p1670577005889579)
> > > 
> > > 
> > > @atzoum Since S3 manager was using different implementation, when we refactored Session config creation in awsutils, we forgot the S3 manager and it introduced the bug so unifying all of them to use same implementation should fix the bug.
> > 
> > 
> > What was the issue this bug caused? Can you please include it in the pull request's title?
> 
> Updated the PR title, let me know if this explains the problem.

From the title I understand that this is not really a bug, but a new feature. Did we ever support role based auth flag in s3 manager?

koladilip commented:
> > > > > what are we fixing and what issue did this cause? [ref](https://rudderlabs.slack.com/archives/C02B31USATA/p1670577005889579)
> > > > 
> > > > 
> > > > @atzoum Since S3 manager was using different implementation, when we refactored Session config creation in awsutils, we forgot the S3 manager and it introduced the bug so unifying all of them to use same implementation should fix the bug.
> > > 
> > > 
> > > What was the issue this bug caused? Can you please include it in the pull request's title?
> > 
> > 
> > Updated the PR title, let me know if this explains the problem.
> 
> From the title I understand that this is not really a bug, but a new feature. Did we ever support role based auth flag in s3 manager?

Actually this is bug as main aws session creator assumes that session config contain rolebased auth flag set properly but s3 manager didn't implement that functionality before so aws session creation is failing for s3 manage when roles are used.

atzoum commented:
> > > > > > what are we fixing and what issue did this cause? [ref](https://rudderlabs.slack.com/archives/C02B31USATA/p1670577005889579)
> > > > > 
> > > > > 
> > > > > @atzoum Since S3 manager was using different implementation, when we refactored Session config creation in awsutils, we forgot the S3 manager and it introduced the bug so unifying all of them to use same implementation should fix the bug.
> > > > 
> > > > 
> > > > What was the issue this bug caused? Can you please include it in the pull request's title?
> > > 
> > > 
> > > Updated the PR title, let me know if this explains the problem.
> > 
> > 
> > From the title I understand that this is not really a bug, but a new feature. Did we ever support role based auth flag in s3 manager?
> 
> Actually this is bug as main aws session creator assumes that session config contain rolebased auth flag set properly but s3 manager didn't implement that functionality before so aws session creation is failing for s3 manage when roles are used.

Is this the PR title then?
`fix: aws session creation is failing for s3 manager when roles are used`

koladilip commented:
> > > > > > > what are we fixing and what issue did this cause? [ref](https://rudderlabs.slack.com/archives/C02B31USATA/p1670577005889579)
> > > > > > 
> > > > > > 
> > > > > > @atzoum Since S3 manager was using different implementation, when we refactored Session config creation in awsutils, we forgot the S3 manager and it introduced the bug so unifying all of them to use same implementation should fix the bug.
> > > > > 
> > > > > 
> > > > > What was the issue this bug caused? Can you please include it in the pull request's title?
> > > > 
> > > > 
> > > > Updated the PR title, let me know if this explains the problem.
> > > 
> > > 
> > > From the title I understand that this is not really a bug, but a new feature. Did we ever support role based auth flag in s3 manager?
> > 
> > 
> > Actually this is bug as main aws session creator assumes that session config contain rolebased auth flag set properly but s3 manager didn't implement that functionality before so aws session creation is failing for s3 manage when roles are used.
> 
> This should be the PR title then? `fix: aws session creation is failing for s3 manager when roles are used`

Updated.

sundernagesh18 commented:
Tested the latest AWS FTR changes on the S3 datalake and the changes are working fine.
